hail/src/main/scala/is/hail/io/fs/AzureStorageFS.scala

```diff
@@ -0,0 +1,388 @@
+        val indexOfFirstSlashAfterPrefix: Int = blobItem.getName.indexOf("/", prefix.length)
+        var blobName = blobItem.getName
+        // if there is a slash after the prefix, then this blob is under a subdirectory
+        if (indexOfFirstSlashAfterPrefix != -1) {
+          blobName = blobItem.getName.substring(0, indexOfFirstSlashAfterPrefix)
+        }
+
+        if (!blobNameSet.contains(blobName)) {
+          blobNameSet = blobNameSet + blobName
+          val blobFileName = s"hail-az://$account/$container/$blobName"
+          statusList += fileStatus(blobFileName)
+        }
+      }
+    }
+    populateStatList()
+
+    statusList.toArray
+  }
+
+  def glob(filename: String): Array[FileStatus] = {
+    var (account, container, path) = getAccountContainerPath(filename)
+    path = dropTrailingSlash(path)
+
+    val components =
+      if (path == "")
+        Array.empty[String]
+      else
+        path.split("/")
+
+    val javaFS = FileSystems.getDefault
+
+    val ab = new mutable.ArrayBuffer[FileStatus]()
+    def f(prefix: String, fs: FileStatus, i: Int): Unit = {
+      assert(!prefix.endsWith("/"), prefix)
+
+      if (i == components.length) {
+        var t = fs
+        if (t == null) {
+          try {
+            t = fileStatus(prefix)
+          } catch {
+            case _: FileNotFoundException =>
+          }
+        }
+        if (t != null)
+          ab += t
+      }
+
+      if (i < components.length) {
+        val c = components(i)
+        if (containsWildcard(c)) {
+          val m = javaFS.getPathMatcher(s"glob:$c")
+          for (cfs <- listStatus(prefix)) {
+            val p = dropTrailingSlash(cfs.getPath)
+            val d = p.drop(prefix.length + 1)
+            if (m.matches(javaFS.getPath(d))) {
+              f(p, cfs, i + 1)
+            }
+          }
+        } else
+          f(s"$prefix/$c", null, i + 1)
+      }
+    }
+
+    f(s"hail-az://$account/$container", null, 0)
+    ab.toArray
+  }
+
+  def fileStatus(filename: String): FileStatus = {
+    val (account, container, path) = getAccountContainerPath(filename)
+
+    if (path == "") {
+      return new BlobStorageFileStatus(s"hail-az://$account/$container", null, 0, true)
+    }
+
+    val blobClient: BlobClient = getBlobClient(filename)
+    val blobContainerClient: BlobContainerClient = getContainerClient(filename)
+
+    val directoryContents = blobContainerClient.listBlobsByHierarchy(dropTrailingSlash(path) + "/")
+    var seenBlobItems: Set[String] = Set()
+    var numSubBlobs: Int = 0
+    val iter = directoryContents.iterator()
+    def countSubBlobs(): Unit = {
+      while (iter.hasNext) {
+        val name = iter.next.getName
+        if (seenBlobItems.contains(name)) return
+        seenBlobItems += name
+        numSubBlobs += 1
+      }
+    }
+    countSubBlobs()
+
+    if (numSubBlobs == 0 && !blobClient.exists()) throw new FileNotFoundException("file isn't found")
+
+    val isDir = numSubBlobs > 0
```

Comment:
If we're only comparing 0 or more than 0, we can just ask if the iterator `hasNext` instead of going through it all.

Reply:
Also, if you look at the python implementation, `is_file` and `is_dir` is determined based on the existence of bob properties. Let's do that instead.

Reply:
I'm not sure if I can implement fileStatus in that way. If you call getProperties on a BlobClient and you get a 404 response code you can deduce that the path you called it on is the path to a directory or that the path does not correspond to a blob in Azure Storage, but you can't distinguish those two outcomes